### ISSUE TITLE
[💡 Feature]: Combine usage of config.exclude and capabilities: { 'wdio:exclude':

### ISSUE THREAD

**Issue by damencho** [webdriverio/webdriverio#13996]
### Is your feature request related to a problem?

Currently, you cannot use both because of:
https://github.com/webdriverio/webdriverio/blob/cef3ee07f23f9029e3a73cfee2ab2326b5d44b05/packages/wdio-config/src/node/ConfigParser.ts#L306

```
// only use capability excludes if (CLI) --exclude or config exclude are not defined
```

### Describe the solution you'd like.

I want to be able to use both and combine the excludes.

### Describe alternatives you've considered.

The use case is when extending a already existing config which has `capabilities: { 'wdio:exclude':` and we want to add some global excludes. The alternative is to go over existing capabilities and add the excludes everywhere (this is for multi remote config).
an

### Additional context

I tried changing the code: 
```
if (Array.isArray(capExclude)) {
    exclude = [ ...exclude, ...ConfigParser.getFilePaths(capExclude, this._config.rootDir, this._pathService)]
}
```
But that breaks 3 tests.

```
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Failed Tests 3 ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯

 FAIL  packages/wdio-config/tests/node/configparser.test.ts > ConfigParser > merge > should overwrite config and capabilities exclude if piped into cli command
AssertionError: expected [] to deeply equal [ Array(1) ]

- Expected
+ Received

- Array [
-   "/Users/dminkov/dev/webdriverio-damencho/packages/wdio-config/tests/__fixtures__/wdio.conf.rdc.ts",
- ]
+ Array []

 ❯ packages/wdio-config/tests/node/configparser.test.ts:466:27
    464|             configParser['merge']({ exclude: [FIXTURES_CONF] })
    465|             const specs = configParser.getSpecs([FIXTURES_CONF, FIXTURES_CONF_RDC], [FIXTURES_CONF_RDC])
    466|             expect(specs).toEqual([FIXTURES_CONF_RDC])
       |                           ^
    467|         })
    468|

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/3]⎯

 FAIL  packages/wdio-config/tests/node/configparser.test.ts > ConfigParser > merge > should overwrite config and capabilities exclude if piped into cli command with suite
AssertionError: expected [] to have a length of 1 but got +0

- Expected
+ Received

- 1
+ 0

 ❯ packages/wdio-config/tests/node/configparser.test.ts:486:27
    484|
    485|             // validate that only the cli exclude is taken into account and the 'configparser' test is not removed
    486|             expect(specs).toHaveLength(1)
       |                           ^
    487|             expect(specs).toContain(configParserPath)
    488|         })

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[2/3]⎯

 FAIL  packages/wdio-config/tests/node/configparser.test.ts > ConfigParser > getSpecs > should exclude/include capability excludes in suites
AssertionError: expected [ Array(1) ] to include '/Users/dminkov/dev/webdriverio-damenc…'
 ❯ packages/wdio-config/tests/node/configparser.test.ts:685:32
    683|             // verify that the capability exclude was ignored since
    684|             // config exclude takes precedence over capability exclude
    685|             expect(getSpecs()).toContain(requireLibPath)
       |                                ^
    686|             expect(getSpecs()).toContain(configParserPath)
    687|

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[3/3]⎯

 Test Files  1 failed | 266 passed | 1 skipped (268)
      Tests  3 failed | 3068 passed | 13 skipped (3084)
   Start at  17:53:19
   Duration  34.75s (transform 9.68s, setup 2.42s, collect 94.33s, tests 97.64s, environment 3.28s, prepare 16.03s)
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by christian-bromann** [comment]
I'ld suggest that the `wdio:exclude` capability , similar to `wdio:specs` should overwrite the root configuration __instead of merging them__. You can still manually merge both lists but this way you can define excludes and specs that only go for a specific capability.

**Comment by damencho** [comment]
> You can still manually merge both lists but this way you can define excludes and specs that only go for a specific capability.

Yeah, I see. What we needed is exactly the opposite. We include with specs/subfolder/** per different capabilities and want to exclude globally. 
I already did the merging. https://github.com/jitsi/jitsi-meet/commit/f238b93f83696a88a33a1aff7a2a077dfb0f4efe#diff-9d0c2f33b90f16b115d6d67e7e69518499bb70fbbed2ccbb8ec62845904fccefR22

What we do is having: 
```
capabilities: {
        participant1: {
            capabilities: {
                browserName: 'chrome',
            }
        },
        participant2: {
            capabilities: {
                browserName: 'chrome',
                'wdio:exclude': [
                    'specs/alone/**'
                ]
            }
        },
        participant3: {
            capabilities: {
                browserName: 'chrome',
                'wdio:exclude': [
                    'specs/alone/**',
                    'specs/2way/**'
                ]
            }
        },
        participant4: {
            capabilities: {
                browserName: 'chrome',
                'wdio:exclude': [
                    'specs/alone/**',
                    'specs/2way/**',
                    'specs/3way/**'
                ]
            }
        }
```
This way, we have suites separated in different folders, each one launched with one, two, three, or four browsers for the test. We are testing multiparty video conferencing, and we need several participants in different scenarios.
And there are limitation in Firefox like not being able to upload files via `browser.uploadFile` or Firefox does not support providing a audio file to be used instead of mic source (chrome has it as `--use-file-for-fake-audio-capture=`) so we need to exclude some tests when overriding the config to be used with Firefox.

As you see we have general rules to include all tests and exclude particular tests.

**Comment by damencho** [comment]
I found a solution for us that works in the commit I referenced in the previous post and understand that you want to keep different logic for the excludes. Should I close the ticket?

**Comment by damencho** [comment]
Seems my solution it is not completely working :( I will continue looking into it today. Any recommendations are welcome.
Thank you.

**Comment by damencho** [comment]
Yeah, so using the wdio:exclude fixes my problem with having the correct number of browsers, but not what specs to be executed and If I use the global exclude that breaks the filtering of the sessions to create ... 
So there is no solution for me :(

**Comment by damencho** [comment]
And there is no option to use just the caps 'wdio:specs' & 'wdio:exclude', you need to have global specs, but you cannot mix the global exclude and the caps 'wdio:exclude'. It is not making sense to me...

**Comment by damencho** [comment]
The only hack I managed to find is checking some custom config in `onWorkerStart` and doing `specs.pop()` on the specs so it is empty and nothing is executed ... and this relay that only one spec is passed to the worker we use.
Any alternatives you can think of?

**Comment by christian-bromann** [comment]
@damencho cycling back on this. Is there anything I can do to help?

**Comment by damencho** [comment]
Well, I don't think there is anything to be done ... I managed to make a [workaround](https://github.com/jitsi/jitsi-meet/pull/15416/commits/18f46671e3f614d6de7f55f806829e0290d96315#diff-c0fe3e2b06f0a27ca9d6bde03ce31b904d70f71f9402663c6eaa728777bb78e1R265), not ideal but works for now ... 

Maybe the docs should state that the global exclude will overwrite the 'wdio:exclude', as I was surprised by that and did not see it in the docs.

**Comment by christian-bromann** [comment]
This is a bug then. The `exclude` arrays should be merged. For example in here:

<img width="748" alt="Screenshot 2025-01-15 at 11 40 28 AM" src="https://github.com/user-attachments/assets/10469871-73b7-49e6-9160-39eff8dc7a85" />

The testrunner should not run any tests. Right now it runs test2.

**Comment by damencho** [comment]
Ok, I can take a look next week. Yep, if I remember correctly that is what I saw too ...

**Comment by damencho** [comment]
But I remember that I tried fixing it and come up to the tests failing as described in the description. So not sure what to do about those.

**Comment by damencho** [comment]
@christian-bromann wdyt about the tests?

**Comment by christian-bromann** [comment]
> wdyt about the tests?

What tests are you referring to?

**Comment by damencho** [comment]
@christian-bromann  I have shared my attempt to fix this, in the description of this ticket, but tests failing stopped me from proceeding.

**Comment by damencho** [comment]
My changed replaced the one at: 
https://github.com/webdriverio/webdriverio/blob/438c790f947e6d127f35c650448916a9a9693579/packages/wdio-config/src/node/ConfigParser.ts#L311

**Comment by christian-bromann** [comment]
> Currently, you cannot use both because of:

I reached the same position in the code. I think there are various scenarios we have to consider here:

- only `exclude` as part of WDIO config
- `exclude` in config + `wdio:exclude` in capability: IMHO `wdio:exclude` should overwrite the config
- `exclude` in config + `wdio:exclude` in capability + `--exclude` as parametet: IMHO `--exclude` should overwrite both other options

The current existing test may not properly reflect this behavior and likely have to be changed as part of this process.

**Comment by damencho** [comment]
What does overwrite mean? Can you give an example?

**Comment by christian-bromann** [comment]
Like in the example above:

```ts
    exclude: [
        'test/specs/test.e2e.ts'
    ],
    capabilities: [{
        browserName: 'chrome',
        'wdio:exclude': ['test/specs/test2.e2e.ts']
    }],
```
the resolved `exclude` property should be `['test/specs/test2.e2e.ts']`.

**Comment by damencho** [comment]
So you combine all available 'wdio:exclude' from the existing capabilities (cause they can be more witl multi remote) and those override the global exclude.

I think, those should be combined, this is what make sense to me. You have a global exclude of `test.e2e.ts` and local `test2.e2e.ts` for the current capability, so that browser chrome should execute these two combined from the global list of tests, not only `test2.e2e.ts`.

For example:

```
specs: [
        'specs/t1.ts',         'specs/t2.ts',         'specs/t3.ts', 
    ],
exclude: [
        'specs/t1.ts'
    ],
capabilities: {
  b1: {
    capabilities: {
      'wdio:exclude': [
         'specs/t2.ts'
        ]
     }
  },
  b2: {
    capabilities: {
      'wdio:exclude': [
         'specs/t3.ts'
        ]
     }
  },
```

In the example above, I think, browser b1 should execute `t3.ts` and browser b2 should execute `t2.ts`.
And the only option of overwrite is when you have a command line arg `--exclude`, that should overwrite the global exclude, because let's say b2 is Edge and the test t3.ts is not suitable for Edge, that's why it is excluded in its capability and if you want to exclude one test from the global list of tests, let's say t2.ts you will overwrite this so you will exclude only t2 and t3 will be executed for Edge browser b2 which is not desired as t3 is not supporting Edge browser case.

**Comment by christian-bromann** [comment]
@damencho you suggestion makes sense. I am ok with merging the exclude values rather than overwriting. Would you be open raising a PR for this?

**Comment by damencho** [comment]
I can tey it these days.

**Comment by damencho** [comment]
@christian-bromann I'm having a hard times with the tests :( 
In the config I have added, I have just:
```
specs: [
        path.join(TEST_ROOT, '**/*.test.ts'),
        path.join(TEST_ROOT, 'node', '*.test.ts')
    ],
    exclude: [],
```
Just that. And I do:
```
            const configParser = await ConfigParserForTest(FIXTURES_MULTI_EXCLUDES_CONF)
            await configParser.initialize()

            console.log('oooo specs', configParser.getSpecs())

```
And it always returns me just:
```
oooo specs [
  '...../dev/webdriverio-damencho/packages/wdio-config/tests/utils.test.ts'
]

```
While there are multiple files there:
```
tests % find . -name "*.test.ts"
./utils.test.ts
./validateConfig.test.ts
./node/FileSystemPathService.test.ts
./node/configparser.test.ts
```

What am I doing wrong? Can you help me?

**Comment by christian-bromann** [comment]
The config parser doesn't use the actual file system and instead uses a [mocked version of it](https://github.com/webdriverio/webdriverio/blob/6ac6835988687426b3d928b6df7b6fe28df81ad9/packages/wdio-config/tests/node/configparser.test.ts#L7-L8). I am getting confused by this as well and haven't added much tests there in the past. If you feel it helps, feel free to write the unit tests in a new file and test the interfaces you need to test without these mocks.

**Comment by christian-bromann** [comment]
I am not sure, properly not :-/

**Comment by damencho** [comment]
And my workaround got broken by this commit :( 
https://github.com/damencho/webdriverio/commit/10fe0271bd3f93e504101c5ea235f1e71c593889#diff-1a50ba617de828487d0c87297c76ad41e0d7e6e459e788fe079dfdb5d31c19cfR324

**Comment by damencho** [comment]
Yeah, I know. I just discovered it and note it to myself here.
The problem is that I cannot figure out a proper way to test this change. So I will probably go on with the change, manually testing and will open a PR.

**Comment by damencho** [comment]
@christian-bromann here it is: https://github.com/webdriverio/webdriverio/pull/14236

**Comment by christian-bromann** [comment]
Fixed in #14236 .. thanks @damencho
